[otlp] Add note on ilogger scopes (#4818)
open-telemetry/opentelemetry-dotnet: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/README.md

for configuring OpenTelemetry with `ILogger` for certain application types such
as ASP.NET Core and .NET Worker.

**ILogger Scopes**: Otlp log exporter supports exporting ILogger scopes.
**Note**: Scope attributes with key set as empty string or `{OriginalFormat}`
are ignored by exporter. Duplicate keys are exported as is.

> 

Reviewer: Couple comments...

* Users have to set `OpenTelemetryLoggerOptions.IncludeScopes = true` to get scopes. Might be good to mention that or link to some other doc going over the options? I found this guide but it could use some improving: https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/docs/logs/customizing-the-sdk#includescopes

* For the "note" use the github style:

   ```markdown
      > **Note**
      > Content goes here
   ```

Author: Updated

for configuring OpenTelemetry with `ILogger` for certain application types such
as ASP.NET Core and .NET Worker.

> **ILogger Scopes**: OTLP Log Exporter supports exporting ILogger scopes as

Reviewer 2: ```suggestion
**ILogger Scopes**: OTLP Log Exporter supports exporting `ILogger` scopes as
```

> **ILogger Scopes**: Otlp log exporter supports exporting ILogger scopes when

Reviewer 3: (Consistency! Check Ln71)

```suggestion
**ILogger Scopes**: OTLP Log Exporter supports exporting ILogger scopes when
```

> **ILogger Scopes**: OTLP Log Exporter supports exporting ILogger scopes when

Reviewer 4: **ILogger Scopes**: OTLP Log Exporter supports exporting ILogger scopes as Attributes. Scopes must be enabled at the 
SDK level using
[IncludeScopes](../../docs/logs/customizing-the-sdk/Readme.md#includescopes)
setting on `OpenTelemetryLoggerOptions`. 

> **Note**
> Scope attributes with key set as empty string or `{OriginalFormat}`
are ignored by exporter. Duplicate keys are exported as is.

^ nit alternate suggestions.

> Attributes. Scopes must be enabled at the SDK level using

Reviewer 2: ```suggestion
attributes. Scopes must be enabled at the SDK level using
```

Note: The single instance returned by `HttpClientFactory` is reused by all
export requests.

> ## Experimental features

Author: @alanwest FYI I took this from your previous PR.

Reviewer 4: categorize this into signals (logs, metrics sub-headings)

The OTLP exporter contains the following experimental features that are

> available only in pre-release package:

Reviewer: ```suggestion
available only in pre-release packages:
```

* The [exemplar
  specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#exemplar)
  is not yet stable. Pre-release builds of the OpenTelemetry .NET SDK and the
  OTLP exporter support sampling and exporting exemplars.
* Support for exporting the following fields on `LogRecord`: `CategoryName`,
  `EventId`, and `Exception`. These fields are unique to OpenTelemetry .NET's
  data model (i.e., they are not reflected in the [standard data
  model](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#log-and-event-record-definition)).

>   Therefore, the conventions for exporting these fields is not yet stable. In

Reviewer 4: ```suggestion
  While they can be exported as attributes, there is no established conventions for the attribute key names. In
```

Reviewer 4: While they can be exported as attributes, there are no established conventions for the attribute key names to be used.

//similar.

Author: updated

Note: The single instance returned by `HttpClientFactory` is reused by all
export requests.

## Experimental features

> The OTLP exporter contains the following experimental features that are

Reviewer 4: anything user need to do to enable experimental feature other than using the pre-release package?

Author: No

available only in pre-release packages:

* The [exemplar
  specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#exemplar)
  is not yet stable. Pre-release builds of the OpenTelemetry .NET SDK and the
  OTLP exporter support sampling and exporting exemplars.
* Support for exporting the following fields on `LogRecord`: `CategoryName`,
  `EventId`, and `Exception`. These fields are unique to OpenTelemetry .NET's
  data model (i.e., they are not reflected in the [standard data
  model](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#log-and-event-record-definition)).
  While they can be exported as attributes, there is no established conventions
  for the attribute key names. In pre-release builds, these fields are exported
  as attributes mapped as follows:
  * `CategoryName` maps to `dotnet.ilogger.category_name`.
  * `EventId` maps to two attributes `Id` and `Name` representing `Event.Id` and

>     `EventId.Name`, respectively.

Reviewer 3: Why is this `EventId.Name` instead of `Event.Name` (it seems super weird comparing with `Event.Id`)?

Author: it is referring to https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging.eventid?view=dotnet-plat-ext-7.0

Reviewer 3: I know but the name we have here sounds wrong.

Author: There was a typo which I fixed. It should be `EventId.Id` not `Event.Id`

Reviewer 3: > There was a typo which I fixed. It should be `EventId.Id` not `Event.Id`. Sorry for the confusion

It's getting even worse.

Author: https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging.eventid.id?view=dotnet-plat-ext-7.0

Reviewer 4: I think @reyang is asking about the "attribute" name?
The attribute names are just "Id" and "Name" as per this doc.
